### ISSUE TITLE
CygWin64 (NetBeans) on Win10 fails to compile the Header

### ISSUE THREAD

**Issue by gurkanctn** [OneLoneCoder/olcPixelGameEngine#4]
I have CygWin64 on Netbeans, on a Win10 PC.

I've tried an hour to solve each problem that I faced on the way. Sharing it to help other Cygwin / Netbeans folks.

**Here's what worked:**
1) replace `_Win32_` --> `__CYGWIN__` (EVERYWHERE inside the Header file!!!)
2) add Compiler option: `-std=c++11`
3) replace `sprintf_s` --> `sprintf`
so that it reads:
`sprintf(sTitle, "OneLoneCoder.com - Pixel Game Engine - %s - FPS: %d", sAppName.c_str(), nFrameCount);
`
4) `#include <locale>`

and **here's where I'm out of solutions:**

error:

> /cygdrive/c/Projects/pxl/olcPixelGameEngine.h:669:(.text+0xae3): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `GdiplusStartup'

    
I tried adding the compiler option: "`-Wl,--image-base -Wl,0x10000000`", which I found on the internet. After this step, I still received several errors, as can be seen below.

I hope to be able to run the PixelGameEngine pretty soon! :)

Thanks!

_______________

```
cd 'C:\Projects\pxl'
C:\cygwin64\bin\make.exe -f Makefile CONF=Debug
"/usr/bin/make" -f nbproject/Makefile-Debug.mk QMAKE= SUBPROJECTS= .build-conf
make[1]: Entering directory '/cygdrive/c/Projects/pxl'
"/usr/bin/make"  -f nbproject/Makefile-Debug.mk dist/Debug/Cygwin-Windows/pxl.exe
make[2]: Entering directory '/cygdrive/c/Projects/pxl'
mkdir -p dist/Debug/Cygwin-Windows
g++ -std=c++11 -lgdiplus    -o dist/Debug/Cygwin-Windows/pxl build/Debug/Cygwin-Windows/olcExampleProgram.o 
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `olc::PixelGameEngine::Start()':
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:669: undefined reference to `GdiplusStartup'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:669:(.text+0xae3): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `GdiplusStartup'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `olc::PixelGameEngine::EngineThread()':
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1165: undefined reference to `__imp_glEnable'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1165:(.text+0x2102): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glEnable'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1166: undefined reference to `__imp_glGenTextures'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1166:(.text+0x2120): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glGenTextures'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1167: undefined reference to `__imp_glBindTexture'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1167:(.text+0x213d): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glBindTexture'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1168: undefined reference to `__imp_glTexParameteri'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1168:(.text+0x2156): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glTexParameteri'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1169: undefined reference to `__imp_glTexParameteri'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1169:(.text+0x216f): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glTexParameteri'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1170: undefined reference to `__imp_glTexEnvf'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1170:(.text+0x218a): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glTexEnvf'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1172: undefined reference to `__imp_glTexImage2D'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1172:(.text+0x21f3): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glTexImage2D'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1300: undefined reference to `__imp_glTexSubImage2D'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1300:(.text+0x26be): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glTexSubImage2D'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1303: undefined reference to `__imp_glBegin'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1303:(.text+0x26cc): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `__imp_glBegin'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1304: undefined reference to `__imp_glTexCoord2f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1304:(.text+0x26e1): additional relocation overflows omitted from the output
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1304: undefined reference to `__imp_glVertex3f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1305: undefined reference to `__imp_glTexCoord2f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1305: undefined reference to `__imp_glVertex3f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1306: undefined reference to `__imp_glTexCoord2f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1306: undefined reference to `__imp_glVertex3f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1307: undefined reference to `__imp_glTexCoord2f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1307: undefined reference to `__imp_glVertex3f'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1308: undefined reference to `__imp_glEnd'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1312: undefined reference to `__imp_SwapBuffers'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1362: undefined reference to `__imp_wglDeleteContext'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `olc::PixelGameEngine::olc_OpenGLCreate()':
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1492: undefined reference to `__imp_ChoosePixelFormat'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1493: undefined reference to `__imp_SetPixelFormat'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1495: undefined reference to `__imp_wglCreateContext'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1496: undefined reference to `__imp_wglMakeCurrent'
/cygdrive/c/Projects/pxl/olcPixelGameEngine.h:1499: undefined reference to `__imp_wglGetProcAddress'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::GdiplusBase::operator new(unsigned long)':
/usr/include/w32api/gdiplus/gdiplusbase.h:38: undefined reference to `GdipAlloc'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::GdiplusBase::operator delete(void*)':
/usr/include/w32api/gdiplus/gdiplusbase.h:46: undefined reference to `GdipFree'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Image::~Image()':
/usr/include/w32api/gdiplus/gdiplusheaders.h:70: undefined reference to `GdipDisposeImage'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Image::~Image()':
/usr/include/w32api/gdiplus/gdiplusheaders.h:70: undefined reference to `GdipDisposeImage'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Image::Clone() const':
/usr/include/w32api/gdiplus/gdiplusheaders.h:75: undefined reference to `GdipCloneImage'
/usr/include/w32api/gdiplus/gdiplusheaders.h:80: undefined reference to `GdipDisposeImage'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Bitmap::Clone() const':
/usr/include/w32api/gdiplus/gdiplusheaders.h:194: undefined reference to `GdipCloneImage'
/usr/include/w32api/gdiplus/gdiplusheaders.h:199: undefined reference to `GdipDisposeImage'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Image::GetHeight()':
/usr/include/w32api/gdiplus/gdiplusimpl.h:153: undefined reference to `GdipGetImageHeight'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Image::GetWidth()':
/usr/include/w32api/gdiplus/gdiplusimpl.h:278: undefined reference to `GdipGetImageWidth'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Bitmap::Bitmap(wchar_t const*, int)':
/usr/include/w32api/gdiplus/gdiplusimpl.h:429: undefined reference to `GdipCreateBitmapFromFileICM'
/usr/include/w32api/gdiplus/gdiplusimpl.h:432: undefined reference to `GdipCreateBitmapFromFile'
build/Debug/Cygwin-Windows/olcExampleProgram.o: In function `Gdiplus::Bitmap::GetPixel(int, int, Gdiplus::Color*) const':
/usr/include/w32api/gdiplus/gdiplusimpl.h:603: undefined reference to `GdipBitmapGetPixel'
collect2: error: ld returned 1 exit status
make[2]: *** [nbproject/Makefile-Debug.mk:63: dist/Debug/Cygwin-Windows/pxl.exe] Error 1
make[2]: Leaving directory '/cygdrive/c/Projects/pxl'
make[1]: *** [nbproject/Makefile-Debug.mk:59: .build-conf] Error 2
make[1]: Leaving directory '/cygdrive/c/Projects/pxl'
make: *** [nbproject/Makefile-impl.mk:40: .build-impl] Error 2

BUİLD FAILED (exit value 2, total time: 1s)

```

**Comment by OneLoneCoder** [comment]
Oh boy 😆 The errors seem to suggest that the opengl and gdiplus libraries are not being linked. This could be that your versions of them have also not been compiled with CygWin, and/or there is a 32/64 bit discrepency? I would have thought treating CygWin like MinGW would have been a better option. The WIN32define I think is necessary for parts of the windows.h file too

I'm curious, why netbeans and cygwin? It seems like a lot of effort to compile Windows applications? I'm not suggesting it isn't legitimate, but is this a common workflow I've just not been exposed too?

**Comment by gurkanctn** [comment]
Hi,

I use Cygwin simply because it was in the tutorial that I had used to set up the C++ toolchain in NetBeans. :)

I will try treating CygWin as MinGW and then I will write my results here. (Right now, I'm trying to rescue data from a broken external drive.)

Thanks for the suggestion.

**Comment by gurkanctn** [comment]
Short update on my side: I haven't been successful with the NetBeans setup. I've switched to Visual Studio 2017, and it ran successfully at the first try!

I'll spend some time enjoying the features of the PixelGameEngine, before retrying my chance with NetBeans. I would prefer VSCode even more, but the VSCode IDE does not even show a console output!

**Comment by OneLoneCoder** [comment]
Hey Gurkanctn, I'm pleased you got things up and running. If you need a reasonably lightweight build environment then I would suggest looking at code::blocks, its a bit old fashioned but works.
I have not really used VS Code, but I gather you can download all manner of plug-ins to make it feel more like an IDE. I suspect that if you have installed visual studio, you can use the compiler toolchain from that, but the code editor from VS code.

**Comment by TheRealNOIG** [comment]
I have started to run into this same issue on CLion for both windows and Linux. Unfortunately, I cannot install VS on Linux. I hope someone can find a solution to this issue

**Comment by TheRealNOIG** [comment]
Ive solved this issue for now by using the 1.1 version of PixelGameEngine
